[BUG] Fix redirect-from Sphinx extension

AnsonTran opened the pull request:
<!--
Thank you so much for your PR!  To help us review your contribution, please check
out the development guide https://matplotlib.org/devdocs/devel/index.html
-->

## PR summary
<!-- Please provide at least 1-2 sentences describing the pull request in detail
(Why is this change required?  What problem does it solve?) and link to relevant
issues and PRs.

Also please summarize the changes in the title, for example "Raise ValueError on
non-numeric input to set_xlim" and avoid non-descriptive titles such as "Addresses
issue #8576".
-->

Closes #27911, closes #21876. This PR fixes the issue of redirect-from directives not
creating the relative refresh pages in the build.

## PR checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A].-->

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [N/A] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [N/A] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [N/A] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [N/A] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

<!--We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.-->

timhoffm reviewed (commented):
The `clear_doc` change to operating on a single document is correct. Clearing everything was a mistake.

Why do we have to delete the consistency checks?

AnsonTran commented:
If the docs build successfully the first time with no changes, Sphinx will load in a cached build environment with ```domain.requests``` already populated. Hence, that check will cause the build to fail

tacaswell commented:
> Sphinx will load in a cached build environment with domain.requests already populated. Hence, that check will cause the build to fail

But that suggests the checks are doing their job and we still have a bug?

AnsonTran commented:
Added the sanity checks back in. Also added a function to clear the domain redirects, so build doesnt fail when rebuilding without doing a ```make clean```.

timhoffm commented on doc/sphinxext/redirect_from.py:
```diff
@@ -73,28 +73,29 @@ def redirects(self):
-            if src not in self.redirects:
-                self.redirects[src] = dst
-            elif self.redirects[src] != dst:
+        for docname in docnames:
+            otherdoc = otherdata['redirects'][docname]
+            if self.redirects[docname]:
+                src, dst = self.redirects[docname]
+                _, otherdst = otherdoc
```
Can you please give the motivation for this change? I find this quite confusing, but I belive this is wrong. AFAICS from l.194 (`domain.redirects[redirected_reldoc] = current_doc`) `self.redirects` is a mapping "old (non-existing) page" --> "new (existing) page". OTOH the parameter `docnames` contains the existing pages. So, I belive it is incorrect to use it as a key in `self.redirects`.

AnsonTran replied:
Hi Tim, the original motivation for the change was to match the implementation here: 
https://github.com/sphinx-doc/sphinx/blob/5523c9bbe4d1415777669330411d4f00ad802f18/sphinx/ext/todo.py#L90-L92

I did make a mistake when adding the sanity check for this function, so I will undo the changes and leave it as is.

QuLogic commented on doc/sphinxext/redirect_from.py:
```diff
@@ -119,3 +118,10 @@ def _generate_redirects(app, exception):
+
+def _clear_redirects(app):
+    domain = app.env.get_domain('redirect_from')
+    if (domain.redirects):
+        domain.redirects.clear()
+        logger.info('clearing cached redirects')
```
Unnecessary parentheses. Also, the log message is written to go first.
```suggestion
    if domain.redirects:
        logger.info('clearing cached redirects')
        domain.redirects.clear()
```

AnsonTran replied:
Implemented, thank you!
